### ISSUE TITLE
WebSocketClientHandshakers are too restrictive on handshake response status

### ISSUE THREAD

**Issue by slandelle** [netty/netty#4528]
Netty 4+

From [RFC6455](https://tools.ietf.org/html/rfc6455#section-4.1):

> If the status code received from the server is not 101, the
       client handles the response per HTTP [RFC2616] procedures.  In
       particular, the client might perform authentication if it
       receives a 401 status code; the server might redirect the client
       using a 3xx status code (but clients are not required to follow
       them), etc.  Otherwise, proceed as follows.

But WebSocketClientHandshakerXX verify crashes with `WebSocketHandshakeException ` when [status isn't 101](https://github.com/netty/netty/blob/netty-4.1.0.Beta8/codec-http/src/main/java/io/netty/handler/codec/http/websocketx/WebSocketClientHandshaker13.java#L215-L217).

This makes Netty WebSockets clients unable to deal with standard HTTP mechanisms such as redirect and non-preemptive authorization.

**Comment by normanmaurer** [comment]
@slandelle couldn't you use HttpClientCodec.decoder() ?

**Comment by slandelle** [comment]
@normanmaurer Sorry, I don't follow.

Here's how I've been doing things prior to 4.1:
* Channel initially have a HttpClientCodec
* Send initial upgrade request
* If status is different from 101, perform standard HTTP operations such as redirect or auth, and perform upgrade request again
* Otherwise, add the WebSocketEncoder and Decoder in the pipeline, based on HttpClientCodec position, and remove HttpClientCodec. HttpClientCodec lookup is based on the name that was used to set it in the first place.

But now with 4.1:
* HttpClient is no longer a Handler in the pipeline. There's an Encoder and a Decoder instead, whose names are not even based to the name that was used to set the HttpClientCodec.
* I had to use the internal names for the Encoder and Decoder (`HttpClientCodec$Decoder#0` and such) to remove them and find out where to place the WebSocket handlers.

[Definitively hacky. Maybe/probably I'm doing it wrong.](https://github.com/AsyncHttpClient/async-http-client/blob/netty4.1/client/src/main/java/org/asynchttpclient/netty/channel/ChannelManager.java#L451-L461)
But I think the new design overlooked the fact that the WebSocket handler didn't cover the fact that the handshake response could be something else than 101 and be perfectly valid and have to be handled downstream.

**Comment by normanmaurer** [comment]
@slandelle so the real problem is more or less that HttpClientCodec is not working as before in the terms that it just "stays" in the pipeline ?

**Comment by slandelle** [comment]
Yes.

**Comment by normanmaurer** [comment]
@slandelle gotcha... I think this is more or less a breaking change so I'm wondering if we should just make it work as before. Does this sound good ?

**Comment by slandelle** [comment]
That would be awesome and make upgrading a bliss!

**Comment by normanmaurer** [comment]
@trustin WDYT ? I can take care of this

**Comment by normanmaurer** [comment]
Also @Scottmitch  :)

**Comment by normanmaurer** [comment]
@trustin @Scottmitch @slandelle another thing we could do is change the `HttpClientCodec` and others to basically still keep itself in the pipeline and remove the encoder/decoder once itself was removed. That said I think maybe we should just remove `ChannelHandlerAppender` and undeprecate `ChannelDuplexHandler`. It just sounds like this one will give the user less surprises as the handler that they add to the pipeline will just stay there and they can remove it as usual. 

I think this is something we need to fix before 4.1.0.Final

**Comment by normanmaurer** [comment]
@trustin ping...

**Comment by trustin** [comment]
Didn't have time to think about this.. As you know duplex handler has a design flaw, so I'm not sure if it's a good idea.

**Comment by normanmaurer** [comment]
@trustin yeah... it has. What you think about my other proposal like keep the 'codec' in the pipeline and ensure the contained encoder/decoder is removed as well if pipeline.remove(codec) is called. This could be done via the handlerRemoved(....) method in the codec itself.

**Comment by normanmaurer** [comment]
I'm working on a fix for this... should be ready soon . Stay tuned :+1:

**Comment by slandelle** [comment]
@normanmaurer I think there's 2 distinct issues:

* WebSocketClientHandshaker shouldn't crash when response status code is different from 101. It should pass the response downstream and the the user retry the handshake later. This way, users could handle redirects, auth schemes that can't be performed preemptively (Kerberos...)
* HttpClientCodec behavior change. What I do w/ Netty 4.0:

```java
pipeline.addAfter(HTTP_CLIENT_CODEC, WS_ENCODER_HANDLER, new WebSocket08FrameEncoder(true));
pipeline.remove(HTTP_CLIENT_CODEC);
pipeline.addBefore(AHC_WS_HANDLER, WS_DECODER_HANDLER, new WebSocket08FrameDecoder(false, false, config.getWebSocketMaxFrameSize()));
pipeline.addAfter(WS_DECODER_HANDLER, WS_FRAME_AGGREGATOR, new WebSocketFrameAggregator(config.getWebSocketMaxBufferSize()));
```

Meaning that I also need pipeline.addAfter to properly work. Or find another way.

**Comment by normanmaurer** [comment]
Yep... First I want to fix the behavior change. Then the rest :)

**Comment by slandelle** [comment]
Thanks. Please ping me when you want me to test.

**Comment by normanmaurer** [comment]
will do

2015-12-18 15:49 GMT+01:00 Stephane Landelle <notifications@github.com>:

**Comment by normanmaurer** [comment]
@slandelle for the restrictive handshake problem would something like be able to validate a FullHttpResponse without throwing exceptions good enough ?

**Comment by slandelle** [comment]
Yes, I think so.

**Comment by schmitch** [comment]
Is this fixed in Netty 4.1.2?

**Comment by normanmaurer** [comment]
@schmitch  nope.. Had no time yet 😞 . But we love contributions ....

**Comment by slandelle** [comment]
@normanmaurer Finally trying to take a stab at this one.
I have a question: what is `WebSocketClientHandshaker#processHandshake` about? It doesn't seem to be used anywhere.
